### ISSUE TITLE
When/why chunkserver is OVERLOADED?

### ISSUE THREAD

**Issue by onlyjob** [moosefs/moosefs#287]
What to tweak in `mfschunkserver.cfg` to control when chunkserver is `OVERLOADED`?

Recently I've commissioned a new super-fast NVMe-based chunkserver and ironically it is in `OVERLOADED` state most of the time. I don't quite understand why -- in the _CGI/Servers_ tab load is usually under 30 yet in _CGI/Resources_ tab the chunkserver is marked `OVERLOADED`.

`nmon` show very little activity on NVMe device and it looks like chunkserver is the bottleneck on a very fast device.

Please advise.

**Comment by acid-maker** [comment]
It is based on WORKERS_MAX parameter. Then there is hysteresis with hi mark set to 3/4 of WORKERS_MAX and lo mark 1/2 of WORKERS_MAX. Whenever number of workers (threads doing I/O jobs) is higher than hi mark then CS send to the master info that it is OVERLOADED, and whenever number of busy workers is lower than lo mark then it sends to the master info that it is no longer OVERLOADED.

Load on the other hand is calculated as number of tasks waiting in the queue for available workers. This is why (especially when CS is in OVERLOADED state) load might be rather small.

Maybe we should redefine load and take into account also number of busy workers? In such case it would be more obvious why the server is OVERLOADED.

I don't think that CS is the bottleneck here - we have tested MFS on cluster of six servers with NVMe drives connected using 100Gb network and we were able to achieve almost full speed (simulataneously on all machines) - the bottleneck we have found there was actually FUSE - one mfsmount was able to read only 4GB/s (about 40% of network speed), we had to run three mounts to get full speed, but I'm sure that bottleneck was FUSE, because using libmfsio we were able to achieve about 9GB/s (about 90% of such network).

BTW. Maybe NVMe drives are fast when you do simple tasks like read or write, but all solid state drives works very differently from mechanical drives. They are not just "faster" - they are "different". For example when you perform a lot of I/O on such SSD drive it can freeze for seconds (usually more than 5 seconds) - probably it needs to perform some self-cleaning tasks from time to time. Solid state drives are much better than mechanical drives when you often read data and rarely write them. For mechanical drives it doesn't matter if you read or write. Likely on SSD we shouldn't use standard file systems (like ext4 or xfs), but specialised file systems (like f2fs) - here we need to perform more test ourselves.

As for now you may try to increase WORKERS_MAX parameter and we will see if it helps.

**Comment by onlyjob** [comment]
Thank you.

> Maybe we should redefine load and take into account also number of busy workers?

That might be a good idea.

> I don't think that CS is the bottleneck here...

Indeed. That's precisely the issue I'm reporting. CS could have performed much better but it is throttled by `OVERLOADED` state. And I tried to double the `WORKERS_MAX` value without any apparent improvement to this behaviour. I'll try to set the value of `WORKERS_MAX` to `1000`.

Pretty sure my particular SSD does not freeze -- it is an Intel P3700. ;)

At the moment I use `ext4` on NVMe. Thank you for idea to try _f2fs_.

**Comment by onlyjob** [comment]
I've made some changes and for few days chunkserver is running with `WORKERS_MAX=1000` on _f2fs_ yet it remains in OVERLOADED state most of the time...

**Comment by acid-maker** [comment]
ok. It looks like your chunkserver is not in overloaded state, but it is wrongly reported as such. My question is:

Where do you see this "OVERLOADED" state? Because when CS is overloaded then in "Servers" tab the load parameter is in brackets. '<','>' means "OVERLOADED", '(',')' means "INTERNAL REBALANCE" and '['.']' also means "OVERLOADED" (old method of overload detection - this might happen when server is overloaded for a longer time).

Likely you see this "OVERLOADED" state in "Resources" tab in storage class. Am I right? Because if yes it has a different meaning.

**Comment by onlyjob** [comment]
Yes in "Resources" tab, as I've said in the very first message... Just one device in the `mfshdd.cfg` so there is no internal rebalancing...

**Comment by acid-maker** [comment]
Ok. I think that this is a result of a bug in ... mfsmaster. Yes, mfschunkserver is ok, it is not overloaded and you can change back WORKERS_MAX to default.

Genesis of this is very old. I'll try to give you a short description.

First time we connected a new chunkserver to our working MFS, we observed that it quickly became a bottle neck of the system. Why? Because system tried to rebalance disk usage and wrote a lot to the new chunkserver. As a consequence, other writes to this new server (client requested ones) were very slow. So we decided we would rather skip this server when we picked the servers to write new client chunks to. To achieve this goal, we switch the new server into "overloaded" state and keep it that way, until it "fills up" with chunks from replication to a certain degree.

This solution worked fine without labels. Unfortunately, with labels, we will have to change it.

I see scenario when you have N servers with label 'A' and all your files are in class '2A'. Then you attach a new server with label 'B'. This is a new server so it will be marked as "filling up" (which is
wrongly reported as OVERLOADED, because internaly it is treated in the same way), but because all your files have storage policy '2A' then no chunks would be replicated to such server and therefore it will stay in this state forever.

I think this is what happened here. First I need to check if it would be easy to distinguish those two states in CLI/CGI and report this state as "FILLING UP" instead of "OVERLOADED". Second I need to think about different way of detecting if server is in such state - somehow take labels into account here. I have some ideas but I need to think about it.

**Comment by onlyjob** [comment]
Thanks, @acid-maker. Your explanation make sense and it looks like I might be facing this very issue.
New label `T` has been made specifically for this chunkserver as it is a first chunkserver on NVMe storage...
